assembly/src/parsers/ast/mod.rs

```diff
@@ -23,19 +27,84 @@ type ProcMap = BTreeMap<String, ProcedureAst>;
 /// Represents a parsed program AST.
 /// This AST can then be furthur processed to generate MAST.
 /// A program has to have a body and no exported procedures.
-#[derive(Debug)]
+#[derive(Debug, Eq, PartialEq)]
 pub struct ProgramAst {
     pub procedures: ProcMap,
     pub body: Vec<Node>,
 }
 
+impl ProgramAst {
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let mut program_ast_bytes = ByteWriter::new();
+
+        // procedures
+        program_ast_bytes.write_u16(self.procedures.len() as u16);
+        for (key, value) in self.procedures.iter() {
+            program_ast_bytes.write_string(key);
+            program_ast_bytes.write_procedure(value);
+        }
+
+        // body
+        program_ast_bytes.write_vec_node(&self.body);
+
+        program_ast_bytes.get_bytes()
+    }
+
+    pub fn from_bytes(bytes: &mut &[u8]) -> Result<Self, SerializationError> {
+        let mut procedures = ProcMap::new();
+
+        let procedures_len = ByteReader::get_u16(bytes).unwrap();
```

Comment:
nit: I'd rename `procedures_len` into `num_procedures`.

Also, we should replace `unwrap()` with `expect()` here and in other places.